File: CodalityTestHarness/CodifyTestHarness/7. Stacks and Queues/Brackets.cs
using System;
using System.Collections.Generic;
using System.Linq;
namespace CodalityTestLibrary
{
    /// <summary>
    /// Write a function that, given a string S consisting of N characters, 
    /// returns 1 if S is properly nested and 0 otherwise.
    /// </summary>
    public class Brackets
    {
        public int solution(string stringToParse)
        {
            // Gaurd against empty
            if(string.IsNullOrEmpty(stringToParse))
            {
                return 1;
            }

            // Guard against un-even char counts
            if(stringToParse.Length % 2 != 0)
            {
                return 0;
            }

            Dictionary<char, char> brackets = new Dictionary<char, char> { { '{', '}' }, { '(', ')' }, { '[', ']' } };
            Stack<char> openedBrackets = new Stack<char>();

            foreach(char character in stringToParse)
            {
                // Add to the stack
                if(brackets.ContainsKey(character))
                {
                    openedBrackets.Push(character);
                }
                // We have encountered a closing bracket or char before 
                // and opening bracket so bail out
                else if(openedBrackets.Count() == 0)
                {
                    return 0;
                }
                // The last opened char added to the stack
                // does not match expected closing bracket
                else if(character != brackets[openedBrackets.Pop()])
                {
                    return 0;
                }
            }

            if(openedBrackets.Count != 0)
            {
                return 0;
            }

            return 1;
        }
    }
}

File: Python/7. Stacks and Queues/Brackets.py
# Write a function that, given a string S consisting of N characters, 
# returns 1 if S is properly nested and 0 otherwise.

class Brackets(object):

    def solution(self, stringToParse):
        
        # guard against empty strings
        if len(stringToParse) == 0:
            return 0

        # guard against uneven strings
        if len(stringToParse) % 2 != 0:
            return 0

        # setup a lookup and stash
        brackets = {"{":"}", "(":")", "[":"]"}
        openedBrackets = []

        for char in stringToParse:
            # check if we have seen this char
            if char in brackets:
                openedBrackets.append(char)
            # if no open brackets then bail
            elif len(openedBrackets) == 0:
                return 0
            # if we don't have the expected closing bracket then also bail
            elif char != brackets[openedBrackets.pop()]:
                return 0

        # at this point we should be bak to an empty stash
        if len(openedBrackets) != 0:
            return 0
        # return happiness
        return 1